zerver/tests/test_subs.py

```diff
@@ -1220,78 +1323,113 @@ def test_cant_remove_others_from_stream(self) -> None:
         """
         result = self.attempt_unsubscribe_of_principal(
-            query_count=21, target_users=[self.example_user('cordelia')], is_admin=True, is_subbed=False,
-            invite_only=True, target_users_subbed=True, other_sub_users=[self.example_user("othello")])
+            query_count=22, target_users=[self.example_user('cordelia')], is_realm_admin=True,
+            is_subbed=False, invite_only=True, target_users_subbed=True, other_sub_users=[self.example_user("othello")])
+        json = self.assert_json_success(result)
+        self.assertEqual(len(json["removed"]), 1)
+        self.assertEqual(len(json["not_removed"]), 0)
+
+    def test_stream_admin_remove_others_from_public_stream(self) -> None:
+        """
+        If you're a stream admin, you can remove people from public streams.
+        """
+        result = self.attempt_unsubscribe_of_principal(
+            query_count=23, target_users=[self.example_user('cordelia')], is_realm_admin=False,
+            is_stream_admin=True, is_subbed=True, invite_only=False, target_users_subbed=True)
+        json = self.assert_json_success(result)
+        self.assertEqual(len(json["removed"]), 1)
+        self.assertEqual(len(json["not_removed"]), 0)
+
+    def test_stream_admin_remove_multiple_users_from_stream(self) -> None:
+        """
+        If you're a stream admin, you can remove multiple users from a stream,
+        """
+        result = self.attempt_unsubscribe_of_principal(
+            query_count=32, target_users=[self.example_user('cordelia'), self.example_user('prospero')],
+            is_realm_admin=False, is_stream_admin=True, is_subbed=True, invite_only=False,
+            target_users_subbed=True)
+        json = self.assert_json_success(result)
+        self.assertEqual(len(json["removed"]), 2)
+        self.assertEqual(len(json["not_removed"]), 0)
+
+    def test_stream_admin_remove_others_from_private_stream(self) -> None:
+        """
+        If you're a stream admin, you can remove people from private streams.
+        """
+        result = self.attempt_unsubscribe_of_principal(
+            query_count=23, target_users=[self.example_user('cordelia')], is_realm_admin=False,
+            is_stream_admin=True, is_subbed=True, invite_only=True, target_users_subbed=True)
         json = self.assert_json_success(result)
         self.assertEqual(len(json["removed"]), 1)
         self.assertEqual(len(json["not_removed"]), 0)
 
     def test_cant_remove_others_from_stream_legacy_emails(self) -> None:
         result = self.attempt_unsubscribe_of_principal(
-            query_count=3, is_admin=False, is_subbed=True, invite_only=False,
-            target_users=[self.example_user('cordelia')], target_users_subbed=True,
+            query_count=6, is_realm_admin=False, is_stream_admin=False, is_subbed=True,
+            invite_only=False, target_users=[self.example_user('cordelia')], target_users_subbed=True,
             using_legacy_emails=True)
```

Comment:
Do you understand why this adds 3 database queries?  I'd have expected just 1 extra query.

Reply:
I did not notice it earlier. But I am also not able to understand why 3 queries are being added.

Reply:
Well, the best way to check is to print out the queries before-and-after and see.

Reply:
The query count bug was fixed by this:

```
-        subs = Subscription.objects.filter(user_profile=user_profile,
-                                           recipient__type=Recipient.STREAM,
-                                           recipient__type_id__in=existing_stream_ids,
-                                           active=True)
+        subs = Subscription.objects.select_related("recipient").filter(
+            user_profile=user_profile,
+            recipient__type=Recipient.STREAM,
+            recipient__type_id__in=existing_stream_ids,
+            active=True)
```